test/variable.jl

```diff
@@ -10,425 +10,513 @@
 # test/variable.jl
 # Testing for VariableRef
 #############################################################################
+
 using JuMP
+
 using Compat
+import Compat.LinearAlgebra: Symmetric
 using Compat.Test
 
-function variables_test(ModelType::Type{<:JuMP.AbstractModel}, VariableRefType::Type{<:JuMP.AbstractVariableRef})
-    AffExprType = JuMP.GenericAffExpr{Float64, VariableRefType}
-
-    @testset "Constructors" begin
-        mcon = ModelType()
+include("utilities.jl")
+@static if !(:JuMPExtension in names(Main))
+    include("JuMPExtension.jl")
```

Comment:
This is a bit sketchy. I think the root of the issue is trying to make one file both includable and executable at the same time?

Reply:
Not doing this is fine too (its warning-only), it just seemed like the most non-hacky way to check if a module is already defined. Its basically an include guard, right down to the compile-time-ness.

Reply:
I'm going to stage new PRs on top of this branch so some time to think about it if you are unsure.

Reply:
Maybe I misunderstood, is the purpose of this include guard just to make it explicit that `variable.jl` depends on `JuMPExtension.jl` or so that the file can be run standalone?

If it's the former, that's not really a pattern that exists in Julia right now. Typically you'd import a module instead?

If it's the latter, how do we test that these test files like `variable.jl` run as standaone files? That's not covered by the regular tests... Wouldn't it be nicer to declare the tests in a file and then have a main `runtests.jl` that decides which subset of tests to run? I guess that leads to your new testing framework :)

Reply:
The purpose is so that variable.jl can be run alone AND as include-from-runtests.jl

To run alone, each file needs to import this JumpExt code.

As runtests.jl will include every .jl file, we'll be running this importing code many times.

I'm not sure how `import` works for modules that are just defined locally in the same directory, and whether calling `import` loads the module multiple times - I can try that.

Running subsets is the goal, but yeah, I think the first step in that direction is just getting the ability to a least run a file by itself / multiple files in parallel.

Reply:
Ok, so how do we test that the tests can run alone?

Reply:
`julia1 --project=. test/variable.jl` works (if you have `Project.toml` created)

Reply:
Oh you want to unit test that tests can be run standalone?
I don't know how to do that for cheap. Certainly possible on Travis but expensive.

Reply:
Is that important? `runtests.jl` will be the "canonical" way, this to be was for developing JuMP and to be a secondary way.

Reply:
> Oh you want to unit test that tests can be run standalone?

Yes, so that this functionality doesn't rot. If we don't test it, we at least need something like a README.md under test/ describing the structure.

Reply:
Can I at least get it working first before getting to that? That was going to be last step.

Reply:
Yes :)